libyara/grammar.y

```diff
@@ -601,40 +617,98 @@ string_declaration
 
 
 string_modifiers
-    : /* empty */                         { $$ = 0; }
-    | string_modifiers string_modifier    { $$ = $1 | $2; }
+    : /* empty */
+      {
+        $$.flags = 0;
+        $$.xor_min = 0;
+        $$.xor_max = 0;
+      }
+    | string_modifiers string_modifier
+      {
+        set_flag_or_error($$.flags, $2.flags);
```

Comment:
This has a problem similar to the one mentioned in my previous comments. The line `set_flag_or_error($$.flags, $2.flags)` does this:

```
$$.flags |= $2.flags;
```

But it turns out that `$$` may have any random value initially (at least in theory). In practice it doesn't have a random value because it gets initialized by Bison as `$$ = $1`, but Bison doesn't guarantee this, it's just an implementation detail that we shouldn't rely on.

Also, my previous comment still holds:

> Yes, that's correct, $1 is the accumulation of all the other modifiers. But $1 could have the XOR flag set and some values in xor_min and xor_max, but if $2 doesn't have the flag, the values already $1 are not copied into $$. The values in $$ are the result from this rule, and could be undefined if not explicitly set.
> 
> It works because the implementation detail in Bison that forces $$ = $1 before calling our code. But this is not documented, in theory if you don't set the result value for your reduction rule in $$ it is undefined.

I mean, this `if` statement is not enough:

```
        if ($2.flags & STRING_GFLAGS_XOR)
        {
          $$.xor_min = $2.xor_min;
          $$.xor_max = $2.xor_max;
        }
```

If the condition is not true `$$.xor_min` and `$$.xor_max` are not initialized, and we could lose any flag or xor limits stored in `$1`. This is not happening for the same reason explained above.

We should do something like:

```
        $$ = $1;   <-- This guarantees that $$ is initialized with whatever we have in $1, even if this already happens in the current implementation of Bison, we must do it explicitly. 

        set_flag_or_error($$.flags, $2.flags);

        if ($2.flags & STRING_GFLAGS_XOR)
        {
          $$.xor_min = $2.xor_min;
          $$.xor_max = $2.xor_max;
        } 
```

Reply:
Thanks for the detailed explanation. I see how it works now, and thank you for pointing it out!